src/components/TimeControl/TimeControlPicker.tsx

```diff
@@ -16,805 +16,309 @@
-        this.syncTimeControl({ total_time: parseInt(ev.target.value) });
-    update_pause_on_weekends = (ev: React.ChangeEvent<HTMLInputElement>) =>
-        this.syncTimeControl({ pause_on_weekends: ev.target.checked });
+    };
 
-    saveSettings() {
-        const speed = this.state.speed;
-        const system = this.state.system;
-        data.set(`time_control.speed`, speed);
-        data.set(`time_control.system`, system);
-        data.set(
-            `time_control.${speed}.${system}`,
-            makeTimeControlParameters(this.state, this.props.boardWidth, this.props.boardHeight),
+    const instantiate = <T extends TimeControl, U extends keyof T & string>(
+        property: U,
+        name: string,
+        id: string,
+        getter: (ev: React.ChangeEvent<HTMLSelectElement>) => T[U],
+        tc: T,
+    ) => {
+        return (
+            <TimeControlPropertySelector<T, U>
+                key={id}
+                tc={tc}
+                id={id}
+                name={name}
+                property={property}
+                valueGetter={getter}
+                onChangeProperty={onChangeProperty}
+            ></TimeControlPropertySelector>
         );
-    }
+    };
 
-    render() {
-        const speed = this.state.speed;
+    let selectors: JSX.Element[] = [];
+    switch (tc.system) {
+        case "fischer":
+            selectors = [
+                instantiate("initial_time", _("Initial Time"), "tc-initial-time", numeric, tc),
+                instantiate("time_increment", _("Time Increment"), "tc-inc-time", numeric, tc),
+                instantiate("max_time", _("Max Time"), "tc-max-time", numeric, tc),
+            ];
+            break;
+        case "simple":
+            selectors = [instantiate("per_move", _("Time per Move"), "tc-per-move", numeric, tc)];
+            break;
+        case "canadian":
+            selectors = [
+                instantiate("main_time", _("Main Time"), "tc-main-time-canadian", numeric, tc),
+                instantiate(
+                    "period_time",
+                    _("Time per Period"),
+                    "tc-per-period-canadian",
+                    numeric,
+                    tc,
+                ),
+                <TimeControlPropertyInput
+                    tc={tc}
+                    key={"tc-stones-per-period"}
+                    id={"tc-stones-per-period"}
+                    name={_("Stones per Period")}
+                    property={"stones_per_period"}
+                    min={default_time_settings[tc.speed].canadian.stones_per_period_min}
+                    max={default_time_settings[tc.speed].canadian.stones_per_period_max}
+                    default={default_time_settings[tc.speed].canadian.stones_per_period}
+                    valueGetter={(ev) => parseInt(ev.target.value)}
+                    onChangeProperty={onChangeProperty}
+                ></TimeControlPropertyInput>,
```

Comment:
Minor nit: `/>` instead of `></TimeControlPropertyInput>`

Reply:
Will change it! Sloppy autocomplete on my part.

We should probably continue to discuss potential time control changes on Slack when you have time, since this rewrite currently only highlights rather than fixes mismatches between user selections and server treatment.